src/sink_http.c

```diff
@@ -156,76 +178,246 @@ static int add_metrics(void* data,
         syslog(LOG_ERR, "Unknown metric type: %d", type);
         break;
     }
+}
+
+
+/*
+ * Callback function for adding a metric into json object. This function
+ * is called by iterator of metrics object.
+ * TODO: There is a lot redundant code here with sink_stream to normalize
+ * an output representation of a metrics.
+ * @arg data The config data, including configs and the output objects.
+ * @arg type The type of metrics.
+ * @arg metric The metirc key.
+ * @arg value The metric value.
+ * @return 0 on success.
+ */
+static int _add_metric_cb(void* data,
+                          metric_type type,
+                          char* metric,
+                          void* value) {
+
+    struct metric_cb_info* info = (struct metric_cb_info*)data;
+    hashmap* json_map = info->json_map;
+    const statsite_config* config = info->config;
+    const sink_config_http* httpconfig = info->httpconfig;
+
+    /* Extract partition info */
+    char* partition;
+    int name_len;
+    char* pos = strchr(metric, PARTITION_DELIMITER);
+    if (pos == NULL) {
+        /* There is no partition data */
+        partition = DEFAULT_PARTITION;
+        name_len = strlen(metric);
+    } else {
+        /* There is partition data, separate name and partition */
+        partition = pos + 1;
+        name_len = pos - metric;
+    }
 
+    /* Early out if the metric name is empty */
+    if (name_len == 0) {
+        return 0;
+    }
+
+    /* Get prefix from config */
+    char* prefix = NULL;
+    uint16_t pre_len = 0;
+    if (httpconfig->use_prefix) {
+        prefix = config->prefixes_final[type];
+        pre_len = strlen(prefix);
+    }
+
+    /*
+     * Find the json object mapped to the partition string.
+     * Create one if there isn't one yet.
+     */
+    json_t* obj = NULL;
+    if (hashmap_get(json_map, partition, (void**)&obj) == -1) {
+        /* Create a new json object if it doesn't exist */
+        obj = json_object();
+        hashmap_put(json_map, partition, obj);
+    }
+
+    assert(obj != NULL);
+    _add_metric_to_json(obj, type, prefix, pre_len, metric, name_len, value, config);
     return 0;
 }
 
-static int json_cb(const char* buf, size_t size, void* d) {
+/*
+ * Callback function for dumping json object to string.
+ * @arg buf The data buffer to dump.
+ * @arg size Number of bytes to dump.
+ * @arg buf The output buffer.
```

Comment:
You meant ```d``` instead of ```buf``` here?

nitpicking: I know you are just trying to rename internal functions here, but just want to point out since you are touching this function anyway- I'm not sure why ```d``` is named as such and of type ```void *```. Seems it'd make more sense to have the signature be more explicit about type assumption, and do the type casting when invoking the function. Also, the original variable name was less than ideal.